flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java

```diff
@@ -71,4 +71,41 @@ public void testKeepFirstRowWithGenerateRetraction() throws Exception {
 		testHarness.close();
 	}
 
+	@Test
+	public void tesKeepFirstRowWithStateTtl() throws Exception {
+		MiniBatchDeduplicateKeepFirstRowFunction func = new MiniBatchDeduplicateKeepFirstRowFunction(typeSerializer, minTime.toMilliseconds());
+		OneInputStreamOperatorTestHarness<BaseRow, BaseRow> testHarness = createTestHarness(func);
+		testHarness.setup();
+		testHarness.open();
+		testHarness.processElement(record("book", 1L, 12));
+		testHarness.processElement(record("book", 2L, 11));
+		// output is empty because bundle not trigger yet.
+		Assert.assertTrue(testHarness.getOutput().isEmpty());
+		testHarness.processElement(record("book", 1L, 13));
+
+		testHarness.setStateTtlProcessingTime(30);
+		//Incremental cleanup is an eventual clean up, more state access guarantee more expired state cleaned
+		for (long i = 3; i < 30; i++) {
+			testHarness.processElement(record("book", i, 20));
```

Comment:
cc @Myasuka , is this a correct way to trigger state cleanup? 
Is there a better to make sure the expired state cleaned?

Reply:
Similar to`TtlStateTestBase#triggerMoreIncrementalCleanupByOtherOps` which is used to trigger the expired state clean eventually.

Reply:
My concern is that why the number is 30? Is 30 enough?

Reply:
In this case, it is enough, I just picked it at random, of course, the bigger the better.

Reply:
Since we choose `ReturnExpiredIfNotCleanedUp` for better performance, and verifying the data has been physically deleted is a topic related to the implementation of clean up strategy for different state backends. For RocksDB, we have to reply on the compaction taken for sst files containing those keys. For Heap KeyedStateBackend, the records to fetch for cleaning up is related to the key groups (how many state map would existed) in one state backend. 
I think state backend module could guarantee the correctness, and this should be out of scope for SQL module. Once state backend changes the implementation of TTL cleanup strategy, verifying the data has been physically deleted in SQL module might be unstable.
In a nutshell, we would not need to ensure data has been physically deleted here, verifying the state in `MiniBatchDeduplicateKeepFirstRowFunction` has been configured with TTL is enough. Otherwise, if we use `NeverReturnExpired` strategy, things would be much simpler here.

Reply:
@Myasuka Thanks for the input.
+1 to use `NeverReturnExpired` strategy. We have got other questions from user ML about this too. IMHO, `NeverReturnExpired` is more straight forward to understand for users.

Reply:
I also discussed with @lincoln-lil and we think `NeverReturnExpired` is better to give a more deterministic result for operators. +1 to use `NeverReturnExpired`. Could you update the strategy @lsyldliu ? And we don't need this foreach then.

Reply:
Hi @libenchao , what's the question in user mailing list?

Reply:
@wuchong a mail titled "关于状态TTL" in user-zh ML.

Reply:
Thanks @libenchao .

Reply:
Okay, I also prefer to `NeverReturnExpired` strategy, I choose `ReturnExpiredIfNotCleanedUp ` because stream-stream join use it currently, I think it should be unified.